tests/base64_tests.cpp

```diff
@@ -0,0 +1,298 @@
+    std::mt19937 gen(std::mt19937::result_type(123456));
+    std::uniform_int_distribution<int> byte_generator{0, 255};
+    for (size_t trial = 0; trial < 10; trial++) {
+      for (size_t i = 0; i < len; i++) {
+        source[i] = byte_generator(gen);
+      }
+      size_t size = implementation.binary_to_base64(
+          source.data(), source.size(), buffer.data());
+      buffer.resize(size);
+      size_t location = add_garbage(buffer, gen);
+      std::vector<char> back(simdutf::maximal_binary_length_from_base64(
+          buffer.data(), buffer.size()));
+      simdutf::result r = simdutf::base64_to_binary(
+          buffer.data(), buffer.size(), back.data());
+      ASSERT_EQUAL(r.error, simdutf::error_code::INVALID_BASE64_CHARACTER);
+      ASSERT_EQUAL(r.count, location);
+    }
+  }
+}
+
+TEST(doomed_truncated_base64_roundtrip) {
+  for (size_t len = 1; len < 2048; len++) {
+    std::vector<char> source(len, 0);
+    std::vector<char> buffer;
+    buffer.resize(implementation.base64_length_from_binary(len));
+    std::mt19937 gen(std::mt19937::result_type(123456));
+    std::uniform_int_distribution<int> byte_generator{0, 255};
+    for (size_t trial = 0; trial < 10; trial++) {
+      for (size_t i = 0; i < len; i++) {
+        source[i] = byte_generator(gen);
+      }
+      size_t size = implementation.binary_to_base64(
+          source.data(), source.size(), buffer.data());
+      buffer.resize(size - 3);
+      std::vector<char> back(simdutf::maximal_binary_length_from_base64(
+          buffer.data(), buffer.size()));
+      simdutf::result r = simdutf::base64_to_binary(
+          buffer.data(), buffer.size(), back.data());
+      ASSERT_EQUAL(r.error, simdutf::error_code::BASE64_INPUT_REMAINDER);
+      ASSERT_EQUAL(r.count, (size-4)/4*3);
+    }
+  }
+}
+
+TEST(roundtrip_base64_with_spaces) {
+  for (size_t len = 0; len < 2048; len++) {
+    std::vector<char> source(len, 0);
+    std::vector<char> buffer;
+    buffer.resize(implementation.base64_length_from_binary(len));
+    std::mt19937 gen(std::mt19937::result_type(123456));
+    std::uniform_int_distribution<int> byte_generator{0, 255};
+    for (size_t trial = 0; trial < 10; trial++) {
+      for (size_t i = 0; i < len; i++) {
+        source[i] = byte_generator(gen);
+      }
+      size_t size = implementation.binary_to_base64(
+          source.data(), source.size(), buffer.data());
+      buffer.resize(size);
+      for (size_t i = 0; i < 5; i++) {
+        add_space(buffer, gen);
+      }
+
+      std::vector<char> back(simdutf::maximal_binary_length_from_base64(
+          buffer.data(), buffer.size()));
+      simdutf::result r = simdutf::base64_to_binary(
+          buffer.data(), buffer.size(), back.data());
+      ASSERT_EQUAL(r.error, simdutf::error_code::SUCCESS);
+
+      back.resize(
+          r.count); // resize the buffer according to actual number of bytes
+      ASSERT_EQUAL(r.count, len);
+      ASSERT_TRUE(back == source);
```

Comment:
We cannot. It will fail with...

```
[ 29%] Building CXX object tests/CMakeFiles/base64_tests.dir/base64_tests.cpp.o
In file included from simdutf/tests/base64_tests.cpp:8:
simdutf/tests/base64_tests.cpp: In function ‘void test_impl_roundtrip_base64(const simdutf::implementation&)’:
/home/dlemire/CVS/github/simdutf/tests/../tests/helpers/test.h:47:32: error: no match for ‘operator<<’ (operand types are ‘std::basic_ostream<char>’ and ‘const std::vector<char>’)
   47 |     std::cout << "\nExpected " << expr << " to be " << b << ".\n";\
      |     ~~~~~~~~~~~~~~~~~~~~~~~~~~ ^~ ~~~~
      |               |                   |
      |               |                   const std::vector<char>
      |               std::basic_ostream<char>
```